homeassistant/components/2n_entrycom/sensor.py

```diff
@@ -0,0 +1,235 @@
+"""Handle 2n Entrycom doorbell."""
+
+import sys
+import asyncio
+import async_timeout
+import aiohttp
+import logging
+import voluptuous as vol
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.entity import Entity
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+
+import homeassistant.helpers.config_validation as cv
+import json
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "2n keycard_validator"
+CONF_USERNAME = "username"
+CONF_PASSWORD = "password"
+CONF_URL = "entrycom_url"
+
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Required(CONF_URL): cv.string,
+    }
+)
+
+
+async def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Load and configure platform."""
+    uname = config.get(CONF_USERNAME)
+    passw = config.get(CONF_PASSWORD)
+    _LOGGER.info("2n setting up")
+    friendly_name = "Doorbell"
+    base_url = "{}api/".format(config.get(CONF_URL))
+
+    sensor = TwoNSensor(uname, passw, friendly_name, base_url, hass)
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, sensor.async_stop_2n_sensor())
+    async_add_devices([sensor], True)
+
+
+class TwoNSensor(Entity):
+    """Representation of a Sensor."""
+
+    def __init__(self, username, password, friendly_name, url, hass):
+        """Initialize the sensor."""
+        _LOGGER.info("initializing 2n_sensor")
+        self._entity_id = "sensor.haustuer"
+        self._name = friendly_name
+        self._uname = username
+        self._passw = password
+        self._base_url = url
+        self._device_class = "2n_sensor"
+        self._hass = hass
+        self._id = None
+        self._state = None
+        self._2n_loop_task = None
+        self._subscribe_url = "{}log/subscribe?filter=KeyPressed,CardEntered".format(
+            self._base_url
+        )
+        self._event_url = "unset"
+        _LOGGER.info("2n setup done")
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Start loop when added."""
+        _LOGGER.info("starting loop for the state of the door")
+        self._2n_loop_task = self._hass.loop.create_task(self.async_get_2n_state())
+
+    @asyncio.coroutine
+    def async_stop_2n_sensor(self):
+        """Stop 2n_loop."""
+        if self._2n_loop_task:
+            self._2n_loop_task.cancel()
+
+    async def async_get_2n_state(self):
+        """Get state of doorbell."""
+        while True:
+            websession = async_get_clientsession(self._hass)
+            _LOGGER.info("2n getting id")
+            self._id = None
+            try:
+                _LOGGER.info("2n getting id2")
+                with async_timeout.timeout(10, loop=self._hass.loop):
+                    response = await websession.get(
```

Comment:
The device specific interface code, I/O and data parsing, should be extracted to a standalone library published on pypi.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#4-communication-with-devices-services

Reply:
yeah, you are right, sorry about that.